src/transformers/generation/utils.py

```diff
@@ -351,6 +359,43 @@ def prepare_inputs_for_generation(self, *args, **kwargs):
             "A model class needs to define a `prepare_inputs_for_generation` method in order to use `.generate()`."
         )
 
+    def set_cache_implementation(self, cache_implementation: Union[CacheImplementation, str], **kwargs):
+        """
+        Simple API to switch cache implementation in a model. Users could also do
+        `model.config.generation_config.cache_implementation = xxx` but they will most likely unexpected
+        behaviour
+
+        Args:
+            cache_implementation (`Union[CacheImplementation, str]`):
+                The target cache implementation
+            kwargs (`dict`, *optional*):
+                Optional key word arguments to be passed. E.g. for SinkCache, it is required to
+                pass `window_length` and `num_sink_tokens`.
+        """
+        if cache_implementation.upper() not in CacheImplementation.__members__:
+            raise ValueError(
+                f"Unrecognized cache implementation - you passed {cache_implementation}. Supported cache implementations are"
+                f" {CacheImplementation.__members__}"
+            )
+
+        if not self._supports_cache_class:
+            raise ValueError(
+                "This model do not currently support switching between cache implementations. Please raise a feature request on GitHub for adding"
+                " the request to support this model: https://github.com/huggingface/transformers"
+            )
+
+        if cache_implementation == CacheImplementation.SINK:
+            if "window_length" not in kwargs and "num_sink_tokens" not in kwargs:
+                raise ValueError(
+                    "You asked to switch to the Sink cache implementation, but you did not pass `window_length` and `num_sink_tokens` to "
+                    "`set_cache_implementation`. Try again with passing these arguments to the method. (e.g. `model.set_cache_implementation('sink', window_length=window_length=508, num_sink_tokens=4)`"
+                )
+            else:
+                self.generation_config.sink_window_length = kwargs.get("window_length")
+                self.generation_config.num_sink_tokens = kwargs.get("num_sink_tokens")
```

Comment:
Just for my own understanding, if the currently set cache implementation is sink, with e.g. `window_length=window_length=508, num_sink_tokens=4`, should I be allowed to call `set_cache_implementation` with updated parameters i.e.

```
config.set_cache_implementation(CacheImplementation.SINK, window_length=1_016, num_sink_tokens=8)
```
?

Reply:
yes I think so !
Per my understanding the way the API is designed currently for sink cache is to pass a `SinkCache()` object to `generate()` through `past_key_values`. If `past_key_values` is already passed in `generate()`, the way I designed things in the PR, we only use the cache from `set_cache_implementation` in case `past_key_values` is not passed to `generate`.

TLDR: If one calls `model.set_cache_implementation(CacheImplementation.SINK)`, they shouldn't call `model.generate(xxx, past_key_values=SinkCache())` --> maybe we should raise a warning saying that they already called `model.set_cache_impl()`, wdyt?